rule-types/github/repo_action_list.yaml

```
@@ -0,0 +1,64 @@
+---
+version: v1
+type: rule-type
+name: repo_action_allow_list
+context:
+  provider: github
+description: |
+  Verifies that the github workflows in a repo only use actions enumerated in the rule.
+guidance: |
+  Having an overview over which actions and reusable workflows are allowed in a repository is important and allows for a better overall security posture.
+
+  For more information, see
+  https://docs.github.com/en/rest/actions/permissions#set-allowed-actions-and-reusable-workflows-for-a-repository
+def:
+  in_entity: repository
+  rule_schema:
+    type: object
+    properties:
+      actions:
+        type: array
+        description: |
+          List of actions that are allowed to be used in the repository.
+          The list should be in the format of `owner/repo`
+          For example, `actions/checkout`
+        items:
+            type: string
+  ingest:
+    type: git
+    git:
+      branch: main
+  eval:
+    type: rego
+    rego:
+      type: constraints
+      violation_format: json
+      def: |
+        package minder
+
+        expected_set := {x | x := input.profile.actions[_]}
+        actions := github_workflow.ls_actions("./.github/workflows")
+
+        violations[{"msg": msg}] {
+          intersection := expected_set & actions
+          not count(intersection) == count(actions)
+
+          difference := [x | x := actions[_]; not intersection[x]]
```

Can this simply be `expected_set - actions`?  If so, you could rewrite this as:

```suggestion
          missing := expected_set - actions
          not count(missing) == 0
          msg = format_message(missing, input.output_format)
```

---

I think we could use a combination of the two. The reason I used the weird list comprehension that creates the `difference` list was that IIRC I had issues formatting a set type later in the formatter. I think the difference you used reads better than the intersection, but I will also test edge cases (empty expected_set or expected_set that contains fewer items than actions). I'd like to use your proposal because it reads better, but I'm not that familiar with rego.

---

What if we merge the implementation as-is and push @evankanderson 's fix in a subsequent PR?

---

in my local testing a variant of what Evan suggested seems to work well. I'll give it a bit to test in staging